extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go

```diff
@@ -135,14 +117,40 @@ func (a *genericActuator) Reconcile(ctx context.Context, worker *extensionsv1alp
 		return err
 	}
 
+	existingMachineDeploymentNames := sets.String{}
+	for _, deployment := range existingMachineDeployments.Items {
+		existingMachineDeploymentNames.Insert(deployment.Name)
+	}
+
 	// Generate machine deployment configuration based on previously computed list of deployments and deploy them.
 	a.logger.Info("Deploying the machine deployments", "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
 	if err := a.deployMachineDeployments(ctx, cluster, worker, existingMachineDeployments, wantedMachineDeployments, workerDelegate.MachineClassKind(), clusterAutoscalerUsed); err != nil {
 		return errors.Wrapf(err, "failed to generate the machine deployment config")
 	}
 
 	// Wait until all generated machine deployments are healthy/available.
-	if err := a.waitUntilWantedMachineDeploymentsAvailable(ctx, cluster, worker, wantedMachineDeployments); err != nil {
+	if err := a.waitUntilWantedMachineDeploymentsAvailable(ctx, cluster, worker, existingMachineDeploymentNames, existingMachineClassNames, wantedMachineDeployments, clusterAutoscalerUsed); err != nil {
+		// check if the machine controller manager is stuck
+		isStuck, msg, err2 := a.IsMachineControllerStuck(ctx, worker)
+		if err2 != nil {
+			a.logger.Error(err2, "failed to check if the Machine Controller Manager is stuck after unsuccessfully waiting for all machine deployments to be ready.", "namespace", worker.Namespace)
+			// continue in order to return `err` and determine error codes
+		}
+
+		if isStuck {
+			podList := corev1.PodList{}
+			if err2 := a.client.List(ctx, &podList, client.InNamespace(worker.Namespace), client.MatchingLabels(map[string]string{"role": "machine-controller-manager"})); client.IgnoreNotFound(err2) != nil {
```

Comment:
You can use `client.MatchingLabels{"role": "machine-controller-manager"}` instead of  `client.MatchingLabels(map[string]string{"role": "machine-controller-manager"})`

```suggestion
			if err2 := a.client.List(ctx, &podList, client.InNamespace(worker.Namespace), client.MatchingLabels{"role": "machine-controller-manager"}); err2 != nil {
```